design/src/docs/anatomy.md

```diff
@@ -547,7 +547,44 @@ pub struct PokemonService<S> {
 ## Plugins
 <!-- TODO(missing_doc): Link to "Write a Plugin" documentation -->
 
-Smithy Rust also provides a way to hook into the upgrade procedure in order to modify the service behavior. This is done via the [`Plugin`](https://github.com/awslabs/smithy-rs/blob/4c5cbc39384f0d949d7693eb87b5853fe72629cd/rust-runtime/aws-smithy-http-server/src/plugin.rs#L31-L41) trait:
+There are a variety of places in which the customer can apply middleware. During the build:
+
+- For a specific operation, for example `GetPokemonSpecies`, the model service can be wrapped by a `Layer` before passing it to `GetPokemonSpecies::from_service` constructor.
+- The `Operation::layer` method can be used to apply a `Layer` to a specific operation _after_ it's been upgraded.
+
+After the build is finalized:
+
+- The entire `PokemonService` HTTP service can be wrapped by a `Layer`.
+- Every `Route` in the `Router` can be wrapped by a `Layer` using `PokemonService::layer`.
+
+Although this provides a reasonably "complete" API, it can be cumbersome in some use cases. Suppose a customer wants to log the operation name when a request is routed to said operation. Writing a `Layer`, `NameLogger`, to log a operation name is simple, however with the current API the customer is forced to do the following
+
+```rust
+let get_pokemon_species = GetPokemonSpecies::from_handler(/* handler */).layer(NameLogger::new("GetPokemonSpecies"));
+let get_storage = GetStorage::from_handler(/* handler */).layer(NameLogger::new("GetStorage"));
+let do_nothing = DoNothing::from_handler(/* handler */).layer(NameLogger::new("DoNothing"));
+/* Repeat for every route... */
+```
+
+Note that `PokemonService::layer` cannot be used here because it applies a _single_ layer uniformly across all `Route`s stored in the `Router`.
```

Comment:
If I understood the machinery correctly, this is not necessarily the problem per se.
Our problem is that the layer we are applying does not access to the `Operation` struct, therefore it has no way to ergonomically retrieve the data that we want in this example (i.e. the name of the operation).

Reply:
Yeah, by the time `PokemonService` exists we've already wrapped them in a `Route` and collected them into some `Router` implementation.

We currently don't hold any additional information other than the model service `S` and the layer `L` in the `Operation`, so if the `Layer` did have access to the specific `Operation` that had been bundled into the `Route` it still wouldn't be able to make any judgements about the name. Furthermore, suppose a `FailOnMissing` had been the source of the `Route`, there's no operation name for it.